### DISCUSSION TITLE
curl DIGEST auth on Windows replaces proxy realm with user domain

### DISCUSSION THREAD

**Discussion by genotrance** [curl/curl#12670] [Q&A]
I am testing DIGEST auth with a proxy using curl on Windows. When I try curl cli as follows:
```
curl --proxy host:port google.com -U domain\user:pass --proxy-digest -v
```
The proxy sends the realm in the `Proxy-Authenticate` header per usual:
```
Proxy-Authenticate: Digest realm="Test", qop="auth", algorithm="MD5", nonce="e39659f84c80ae751608aed2f2795117", opaque="ba4238fce21364059d8007a9f156f6c6"
```

Curl responds with a `Proxy-Authorization` header that replaces the realm `Test` from the proxy with the `domain` of the user:
```
Proxy-Authorization: Digest username="user",realm="domain",nonce="28f64b73ad70fa2c04c2423beb3ecda1",uri="/",cnonce="22fbbb7c4af10dcba580e9cb6b865987",nc=00000001,algorithm=MD5,response="11245113a01622b08e9d398517c8743f",qop="auth",opaque="192221dc95a8ec1bdf66260d275cfa50"
```
When I try the same on Linux, the realm is retained and `username="domain\user"` as expected. Any ideas what's going on?

**Comment by genotrance** [comment]
Note also that when I set `-U "Test\domain\user:pass"`, it works fine.

**Comment by jay** [comment]
> Curl responds with a `Proxy-Authorization` header that replaces the realm `Test` from the proxy with the `domain` of the user:
> 
> ```
> Proxy-Authorization: Digest username="user",realm="domain",nonce="28f64b73ad70fa2c04c2423beb3ecda1",uri="/",cnonce="22fbbb7c4af10dcba580e9cb6b865987",nc=00000001,algorithm=MD5,response="11245113a01622b08e9d398517c8743f",qop="auth",opaque="192221dc95a8ec1bdf66260d275cfa50"
> ```
> 
> When I try the same on Linux, the realm is retained and `username="domain\user"` as expected. Any ideas what's going on?

Probably you are using a curl build that uses a digest response generated by Windows OS native SSPI functions. Run `curl -V` to and look for SSPI in the features list. Windows seems to set the realm based on the input of [SEC_WINNT_AUTH_IDENTITY](https://learn.microsoft.com/en-us/windows/win32/api/rpcdce/ns-rpcdce-sec_winnt_auth_identity_a) member 'Domain' but domain is [not documented that way](https://learn.microsoft.com/en-us/windows/win32/secauthn/obtaining-alternate-digest-credentials). curl will interfere with a function Curl_override_sspi_http_realm: _"The realm is extracted from the challenge message and used as the domain if it is not already explicitly set."_ Since the domain is already set it is not overridden.

https://github.com/curl/curl/blob/7161cb17c01dcff1dc5bf89a18437d9d729f1ecd/lib/vauth/digest_sspi.c#L489-L505

https://github.com/curl/curl/blob/7161cb17c01dcff1dc5bf89a18437d9d729f1ecd/lib/vauth/digest_sspi.c#L236-L305

Since you are the only person to bring this up in many years I'd guess it's working as intended but am not certain.

/cc @captain-caveman2k @DigitalDJ

**Reply by genotrance** [reply]
Appreciate the details. I have no idea what's right frankly so will let things be for now. I'd have thought the realm was determined by the server and not the client but who knows.
